File: php/testcode/xxe_021.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_xxe_xslt_user_stylesheet
function xxe021(BenchmarkRequest $req): BenchmarkResponse {
    $xslPath = $req->param('xsl');
    $xsl = new DOMDocument();
    $xsl->load($xslPath); // vuln-code-snippet vuln-line php_xxe_xslt_user_stylesheet
    $proc = new XSLTProcessor();
    $proc->importStylesheet($xsl);
    return BenchmarkResponse::ok('imported');
}
// vuln-code-snippet end php_xxe_xslt_user_stylesheet

File: ruby/testcode/xxe_021.rb
require_relative 'shared'
require 'nokogiri'

# vuln-code-snippet start ruby_xxe_nokogiri_dtdload
def parse_dtdload_handler(req)
  xml = req.post('xml')
  doc = Nokogiri::XML.parse(xml) { |c| c.dtdload } # vuln-code-snippet vuln-line ruby_xxe_nokogiri_dtdload
  BenchmarkResponse.json({ root: doc.root&.name })
end
# vuln-code-snippet end ruby_xxe_nokogiri_dtdload
